Handle subscription list for specific topic separately.

andrazhribernik opened the pull request.

andrazhribernik commented:
I signed it!

jgeewax commented:
Not sure I understand what's happening here. Can you share a bit more about what you're trying to do? And maybe add some tests?

andrazhribernik commented:
Sorry for not adding any description and tests. It is basically a bug fix.

When we tried to use this functionality:

    from gcloud import pubsub
    client = pubsub.Client(project=<project_name>)
    client.list_subscriptions(topic_name=<topic_name>)

we got the following error.

```
TypeError                                 Traceback (most recent call last)
<ipython-input-3-8be845bada22> in <module>()
----> 1 client.list_subscriptions(topic_name="test_aw_2")

/Users/andrazhribernik/cytora/gcloud-python/gcloud/pubsub/client.py in list_subscriptions(self, page_size, page_token, topic_name)
    129         subscriptions = [Subscription.from_api_repr(resource, self,
    130                                                     topics=topics)
--> 131                          for resource in resp['subscriptions']]
    132         return subscriptions, resp.get('nextPageToken')
    133 

/Users/andrazhribernik/cytora/gcloud-python/gcloud/pubsub/subscription.pyc in from_api_repr(cls, resource, client, topics)
     66         if topics is None:
     67             topics = {}
---> 68         topic_path = resource['topic']
     69         topic = topics.get(topic_path)
     70         if topic is None:

TypeError: string indices must be integers
```

It also fails with different error if param topic_name is not specified

    from gcloud import pubsub
    client = pubsub.Client(project=<project_name>)
    client.list_subscriptions()

returns
```
ValueError                                Traceback (most recent call last)
<ipython-input-4-39df795327d3> in <module>()
----> 1 client.list_subscriptions()

/Users/andrazhribernik/cytora/gcloud-python/gcloud/pubsub/client.py in list_subscriptions(self, page_size, page_token, topic_name)
    129         subscriptions = [Subscription.from_api_repr(resource, self,
    130                                                     topics=topics)
--> 131                          for resource in resp['subscriptions']]
    132         return subscriptions, resp.get('nextPageToken')
    133 

/Users/andrazhribernik/cytora/gcloud-python/gcloud/pubsub/subscription.pyc in from_api_repr(cls, resource, client, topics)
     71             # NOTE: This duplicates behavior from Topic.from_api_repr to avoid
     72             #       an import cycle.
---> 73             topic_name = topic_name_from_path(topic_path, client.project)
     74             topic = topics[topic_path] = client.topic(topic_name)
     75         _, _, _, name = resource['name'].split('/')

/Users/andrazhribernik/cytora/gcloud-python/gcloud/pubsub/_helpers.pyc in topic_name_from_path(path, project)
     36     if (len(path_parts) != 4 or path_parts[0] != 'projects' or
     37             path_parts[2] != 'topics'):
---> 38         raise ValueError('Expected path to be of the form '
     39                          'projects/{project}/topics/{topic_name}')
     40     if (len(path_parts) != 4 or path_parts[0] != 'projects' or

ValueError: Expected path to be of the form projects/{project}/topics/{topic_name}
```

The code we implemented basically solve a problem with inconsistency of GCloud REST API. It worked for us and was very quick fix. If you think this should be solved in a different way, I would be happy to hear your ideas. 

I just took a look on tests now and it seems that [tests](https://github.com/GoogleCloudPlatform/gcloud-python/blob/master/gcloud/pubsub/test_client.py#L85-L155) do not reflect the output from GCloud REST API.

GCloud REST API response for a specific topic does not return a list of format [{'name': SUB_PATH, 'topic': TOPIC_PATH}], but returns list of strings (['topic1 name', 'topic2 name', ...]). It also seems pagination is on by default, which is not reflected in tests. 

I am wondering. Is there a bug on GCLOUD REST API or on the gcloud-python side. If you think that this needs to be fix, I can refactor tests as well. 

I hope it is clearer now. If you need any more informations, please do not hesitate to ask me.

jgeewax commented:
/cc @tseaver : This seems like a legit concern. Curious why our system tests don't catch this?

tseaver commented on setup.py:
```diff
@@ -22,7 +22,7 @@
 setup(
     name='gcloud',
-    version='0.8.0',
+    version='0.8.1',
```
Please don't bump release numbers in the PR:  we will handle that when actually making a release.

tseaver commented:
@andrazhribernik can you supply the value of `resource` when the tracebacks are raised?  I'm having a hard time following the logic for your changes.

andrazhribernik commented:
@tseaver Value of resource is different whether param 'topic_name' is specified or not. 

1. `topic_name` is not specified
```
from gcloud import pubsub
client = pubsub.Client(project=<project_name>)
client.list_subscriptions()
```
Error is raised, if there is any subscription for a topic that has already been deleted (`_deleted-topic_`).
Sample value of `resp` object:
```
{u'subscriptions': [
     {
        u'topic': u'projects/<project_id>/topics/<topic_name>',
        u'ackDeadlineSeconds': 10,
        u'pushConfig': {},
        u'name': u'projects//<project_id>/subscriptions//<subscription_name>''
    },
    {
        u'topic': u'_deleted-topic_',
        u'ackDeadlineSeconds': 10,
        u'pushConfig': {},
        u'name': u'projects/<project_id>/subscriptions/<subscription_name>'
    },
```

2. `topic_name` is specified
```
from gcloud import pubsub
client = pubsub.Client(project=<project_name>)
client.list_subscriptions(topic_name=<topic_name>)
```
Error is raised for every request of this type.
Sample value of `resp` object:
```
{u'subscriptions': [
        u'projects/<project_id>/subscriptions/<subscription_name1>',
        u'projects/<project_id>/subscriptions/<subscription_name2>'
    ]
}
```

I hope this explanation will help to understand my code updates changes.

tseaver commented:
@andrazhribernik  I believe that #1440 should address this issue.  @jgeewax, it also adds a system test for `list_subscriptions` (adding one for the `list_topics` case is too hard, as there is no guarantee that a given project will have no topics at all).
